File: taikyoku_engine/board.py
"""Board representation for Taikyoku Shogi."""

from .pieces import (
    BOARD_SIZE, BLACK, WHITE, MOVEMENTS, PROMOTES_TO,
    ROYAL_PIECES, PIECE_RANK, RANK_ROYAL, get_initial_board,
    PIECE_NAME, PIECE_VALUE,
)
from .move import Move


class TaikyokuBoard:
    """36x36 board for Taikyoku Shogi.

    Board is stored as board[row][col] where:
      - Row 0 = top (White's back rank)
      - Row 35 = bottom (Black's back rank)
      - Each cell is (piece_abbrev, color) or None
    """

    def __init__(self):
        self.board = [[None] * BOARD_SIZE for _ in range(BOARD_SIZE)]
        self.side_to_move = BLACK
        self.move_number = 1
        self.move_history = []  # Stack of (move, undo_info) for undo
        # Track royal pieces for each side
        self.royals = {BLACK: set(), WHITE: set()}
        # Track all piece positions for each side
        self.piece_positions = {BLACK: {}, WHITE: {}}  # color -> {(r,c): abbrev}

    def setup_initial(self):
        """Set up the initial position."""
        self.board = get_initial_board()
        self.side_to_move = BLACK
        self.move_number = 1
        self.move_history = []
        self._rebuild_indices()

    def _rebuild_indices(self):
        """Rebuild piece position indices from board state."""
        self.royals = {BLACK: set(), WHITE: set()}
        self.piece_positions = {BLACK: {}, WHITE: {}}
        for r in range(BOARD_SIZE):
            for c in range(BOARD_SIZE):
                cell = self.board[r][c]
                if cell is not None:
                    piece, color = cell
                    self.piece_positions[color][(r, c)] = piece
                    if piece in ROYAL_PIECES:
                        self.royals[color].add((r, c))

    def at(self, row, col):
        """Get piece at (row, col). Returns (abbrev, color) or None."""
        return self.board[row][col]

    def set_piece(self, row, col, piece, color):
        """Place a piece on the board."""
        old = self.board[row][col]
        if old is not None:
            old_piece, old_color = old
            self.piece_positions[old_color].pop((row, col), None)
            if old_piece in ROYAL_PIECES:
                self.royals[old_color].discard((row, col))

        self.board[row][col] = (piece, color)
        self.piece_positions[color][(row, col)] = piece
        if piece in ROYAL_PIECES:
            self.royals[color].add((row, col))

    def remove_piece(self, row, col):
        """Remove piece from (row, col)."""
        cell = self.board[row][col]
        if cell is not None:
            piece, color = cell
            self.piece_positions[color].pop((row, col), None)
            if piece in ROYAL_PIECES:
                self.royals[color].discard((row, col))
        self.board[row][col] = None

    def apply_move(self, move: Move):
        """Apply a move to the board. Returns undo info."""
        fr, fc = move.from_sq
        tr, tc = move.to_sq
        piece, color = self.board[fr][fc]

        # Save undo info
        undo = {
            'from_cell': self.board[fr][fc],
            'to_cell': self.board[tr][tc],
            'side': self.side_to_move,
            'move_number': self.move_number,
        }

        # Handle range captures (capture all pieces along path)
        if move.range_captures:
            undo['range_captures'] = []
            for rr, rc, rp, rcol in move.range_captures:
                undo['range_captures'].append((rr, rc, self.board[rr][rc]))
                self.remove_piece(rr, rc)

        # Handle lion mid-capture
        if move.mid_sq is not None:
            mr, mc = move.mid_sq
            undo['mid_cell'] = self.board[mr][mc]
            self.remove_piece(mr, mc)

        # Handle igui (capture without moving)
        if move.is_igui:
            captured_cell = self.board[tr][tc]
            undo['to_cell'] = captured_cell
            self.remove_piece(tr, tc)
            # Piece stays in place but may promote
            if move.promotion and piece in PROMOTES_TO:
                new_piece = PROMOTES_TO[piece]
                self.remove_piece(fr, fc)
                self.set_piece(fr, fc, new_piece, color)
            self.side_to_move = 1 - self.side_to_move
            if self.side_to_move == BLACK:
                self.move_number += 1
            self.move_history.append((move, undo))
            return undo

        # Standard move
        # Remove piece from origin
        self.remove_piece(fr, fc)

        # Capture at destination
        if self.board[tr][tc] is not None:
            self.remove_piece(tr, tc)

        # Determine if piece promotes
        final_piece = piece
        if move.promotion and piece in PROMOTES_TO:
            final_piece = PROMOTES_TO[piece]

        # Place piece at destination
        self.set_piece(tr, tc, final_piece, color)

        # Update side to move
        self.side_to_move = 1 - self.side_to_move
        if self.side_to_move == BLACK:
            self.move_number += 1

        self.move_history.append((move, undo))
        return undo

    def undo_move(self):
        """Undo the last move."""
        if not self.move_history:
            return

        move, undo = self.move_history.pop()
        fr, fc = move.from_sq
        tr, tc = move.to_sq

        # Restore side to move and move number
        self.side_to_move = undo['side']
        self.move_number = undo['move_number']

        if move.is_igui:
            # Restore the piece at from (may have promoted)
            self.remove_piece(fr, fc)
            from_piece, from_color = undo['from_cell']
            self.set_piece(fr, fc, from_piece, from_color)
            # Restore captured piece at to
            if undo['to_cell'] is not None:
                tp, tc_col = undo['to_cell']
                self.set_piece(tr, tc, tp, tc_col)
            return

        # Remove the moved piece from destination
        self.remove_piece(tr, tc)

        # Restore origin
        if undo['from_cell'] is not None:
            fp, fcol = undo['from_cell']
            self.set_piece(fr, fc, fp, fcol)

        # Restore destination (captured piece)
        if undo['to_cell'] is not None:
            tp, tcol = undo['to_cell']
            self.set_piece(tr, tc, tp, tcol)

        # Restore mid-capture
        if move.mid_sq is not None and 'mid_cell' in undo:
            mr, mc = move.mid_sq
            if undo['mid_cell'] is not None:
                mp, mcol = undo['mid_cell']
                self.set_piece(mr, mc, mp, mcol)

        # Restore range captures
        if 'range_captures' in undo:
            for rr, rc, cell in undo['range_captures']:
                if cell is not None:
                    rp, rcol = cell
                    self.set_piece(rr, rc, rp, rcol)

    def is_valid(self, row, col):
        """Check if (row, col) is within the board."""
        return 0 <= row < BOARD_SIZE and 0 <= col < BOARD_SIZE

    def has_royals(self, color):
        """Check if the given side still has royal pieces."""
        return len(self.royals[color]) > 0

    def get_game_result(self):
        """Check game result.

        Returns:
          'black_wins' if White has no royals
          'white_wins' if Black has no royals
          'draw' if neither has royals (unlikely)
          None if game is ongoing
        """
        black_has = self.has_royals(BLACK)
        white_has = self.has_royals(WHITE)
        if not white_has and not black_has:
            return 'draw'
        if not white_has:
            return 'black_wins'
        if not black_has:
            return 'white_wins'
        return None

    def display(self, compact=True):
        """Return a string representation of the board."""
        lines = []
        for r in range(BOARD_SIZE):
            row_str = []
            for c in range(BOARD_SIZE):
                cell = self.board[r][c]
                if cell is None:
                    row_str.append('.. ')
                else:
                    piece, color = cell
                    prefix = 'v' if color == WHITE else '^'
                    abbrev = piece[:2].ljust(2)
                    row_str.append(f"{prefix}{abbrev}")
            lines.append(f"{36-r:2d} {''.join(row_str)}")
        return '\n'.join(lines)

    def copy(self):
        """Create a deep copy of the board."""
        new = TaikyokuBoard()
        new.board = [row[:] for row in self.board]
        new.side_to_move = self.side_to_move
        new.move_number = self.move_number
        new._rebuild_indices()
        return new

File: src/board.rs
use crate::types::*;
use crate::pieces;

pub struct Board {
    pub cells: [Cell; NUM_SQUARES],
    pub side_to_move: u8,
    pub move_number: u32,
    // Piece lists per color
    pub piece_list: [[u16; MAX_PIECES_PER_SIDE]; 2],
    pub piece_list_len: [usize; 2],
    pub piece_count: [usize; 2],
    // Royal piece tracking
    pub royal_list: [[u16; MAX_ROYALS]; 2],
    pub royal_count: [usize; 2],
    // Undo stack
    history: Vec<UndoInfo>,
}

impl Clone for Board {
    fn clone(&self) -> Self {
        Board {
            cells: self.cells,
            side_to_move: self.side_to_move,
            move_number: self.move_number,
            piece_list: self.piece_list,
            piece_list_len: self.piece_list_len,
            piece_count: self.piece_count,
            royal_list: self.royal_list,
            royal_count: self.royal_count,
            history: self.history.clone(),
        }
    }
}

impl Board {
    pub fn new() -> Self {
        Board {
            cells: [EMPTY_CELL; NUM_SQUARES],
            side_to_move: BLACK,
            move_number: 1,
            piece_list: [[INVALID_SQ; MAX_PIECES_PER_SIDE]; 2],
            piece_list_len: [0; 2],
            piece_count: [0; 2],
            royal_list: [[INVALID_SQ; MAX_ROYALS]; 2],
            royal_count: [0; 2],
            history: Vec::new(),
        }
    }

    pub fn setup_initial(&mut self) {
        self.cells = [EMPTY_CELL; NUM_SQUARES];
        self.side_to_move = BLACK;
        self.move_number = 1;
        self.piece_list = [[INVALID_SQ; MAX_PIECES_PER_SIDE]; 2];
        self.piece_list_len = [0; 2];
        self.piece_count = [0; 2];
        self.royal_list = [[INVALID_SQ; MAX_ROYALS]; 2];
        self.royal_count = [0; 2];
        self.history.clear();

        // Place Black's pieces (rows 24-35)
        for (rank_idx, rank_str) in pieces::SETUP_RANKS.iter().enumerate() {
            let rank_pieces = pieces::parse_setup_rank(rank_str);
            let row = 35 - rank_idx; // Rank 1 at row 35
            for (col, piece_opt) in rank_pieces.iter().enumerate() {
                if let Some(pt) = piece_opt {
                    self.place_piece(row, col, *pt, BLACK);
                }
            }
        }

        // Place White's pieces (rows 0-11) — 180-degree rotation
        for (rank_idx, rank_str) in pieces::SETUP_RANKS.iter().enumerate() {
            let rank_pieces = pieces::parse_setup_rank(rank_str);
            let row = rank_idx; // Rank 1 at row 0
            let len = rank_pieces.len();
            for (col, piece_opt) in rank_pieces.iter().enumerate() {
                if let Some(pt) = piece_opt {
                    self.place_piece(row, len - 1 - col, *pt, WHITE);
                }
            }
        }
    }

    fn place_piece(&mut self, row: usize, col: usize, pt: u16, color: u8) {
        let sq = sq_index(row, col);
        self.cells[sq] = make_cell(pt, color);
        let c = color as usize;

        // Add to piece list
        let idx = self.piece_list_len[c];
        self.piece_list[c][idx] = sq as u16;
        self.piece_list_len[c] = idx + 1;
        self.piece_count[c] += 1;

        // Track royals
        if pieces::is_royal(pt) {
            let ri = self.royal_count[c];
            self.royal_list[c][ri] = sq as u16;
            self.royal_count[c] = ri + 1;
        }
    }

    /// Get the cell value at a given square index.
    #[inline]
    #[allow(dead_code)]
    pub fn at(&self, sq: usize) -> Cell {
        self.cells[sq]
    }

    pub fn apply_move(&mut self, m: &Move) {
        let from = m.from_sq as usize;
        let to = m.to_sq as usize;
        let from_cell = self.cells[from];
        let to_cell = self.cells[to];
        let pt = cell_piece(from_cell);
        let color = cell_color(from_cell);
        let c = color as usize;
        let _opp = 1 - c;

        let mut undo = UndoInfo {
            from_sq: m.from_sq, to_sq: m.to_sq,
            from_cell, to_cell,
            side: self.side_to_move,
            move_number: self.move_number,
            mid_sq: m.mid_sq, mid_cell: EMPTY_CELL,
            range_caps: None,
        };

        // Handle range captures
        if let Some(ref caps) = m.range_caps {
            let mut saved = Vec::new();
            for &(sq, _cap_pt, _cap_color) in caps {
                saved.push((sq, self.cells[sq as usize]));
                self.remove_from_lists(sq as usize);
                self.cells[sq as usize] = EMPTY_CELL;
            }
            undo.range_caps = Some(saved);
        }

        // Handle lion mid-capture
        if m.mid_sq != INVALID_SQ {
            let msq = m.mid_sq as usize;
            undo.mid_cell = self.cells[msq];
            self.remove_from_lists(msq);
            self.cells[msq] = EMPTY_CELL;
        }

        // Handle igui
        if m.is_igui {
            if to_cell != EMPTY_CELL {
                self.remove_from_lists(to);
                self.cells[to] = EMPTY_CELL;
            }
            // Piece may promote in place
            if m.promotion {
                if let Some(promo_pt) = pieces::promotes_to(pt) {
                    let new_cell = make_cell(promo_pt, color);
                    self.cells[from] = new_cell;
                    self.update_royal_status(from, pt, promo_pt, c);
                }
            }
            self.side_to_move = 1 - self.side_to_move;
            if self.side_to_move == BLACK { self.move_number += 1; }
            self.history.push(undo);
            return;
        }

        // Remove from origin
        self.cells[from] = EMPTY_CELL;
        self.remove_sq_from_piece_list(from, c);
        if pieces::is_royal(pt) {
            self.remove_sq_from_royal_list(from, c);
        }

        // Capture at destination
        if to_cell != EMPTY_CELL {
            self.remove_from_lists(to);
        }

        // Determine final piece (promotion)
        let final_pt = if m.promotion {
            pieces::promotes_to(pt).unwrap_or(pt)
        } else {
            pt
        };

        // Place at destination
        self.cells[to] = make_cell(final_pt, color);
        self.add_sq_to_piece_list(to, c);
        if pieces::is_royal(final_pt) {
            let ri = self.royal_count[c];
            self.royal_list[c][ri] = to as u16;
            self.royal_count[c] = ri + 1;
        }

        self.side_to_move = 1 - self.side_to_move;
        if self.side_to_move == BLACK { self.move_number += 1; }
        self.history.push(undo);
    }

    pub fn undo_move(&mut self) -> bool {
        let undo = match self.history.pop() {
            Some(u) => u,
            None => return false,
        };

        self.side_to_move = undo.side;
        self.move_number = undo.move_number;

        let from = undo.from_sq as usize;
        let to = undo.to_sq as usize;

        // Check if it was igui
        if undo.from_sq == undo.to_sq && undo.to_cell != EMPTY_CELL {
            // Restore piece at from to original
            self.cells[from] = undo.from_cell;
            // Restore captured piece at to (same square, but it was an adjacent capture)
            // Actually for igui, from == to but the captured piece was adjacent
            // Let me re-check the igui logic...
            // For igui: the piece stays at from, and an adjacent piece at some other square was captured.
            // But in our Move encoding, to_sq == from_sq for igui and the captured piece is tracked separately.
            // The undo needs to restore from_cell and to_cell.
            self.cells[from] = undo.from_cell;
            // to_cell was the captured piece... but to == from for igui
            // Actually let me look at how apply_move handles igui:
            // It removes to_cell (the adjacent captured piece) from lists, sets cells[to] = EMPTY
            // But to == from in igui... that's a problem in our encoding.
            // For simplicity, let's just rebuild lists.
            self.rebuild_lists();
            // Restore mid/range captures
            if undo.mid_sq != INVALID_SQ {
                self.cells[undo.mid_sq as usize] = undo.mid_cell;
            }
            if let Some(ref caps) = undo.range_caps {
                for &(sq, cell) in caps {
                    self.cells[sq as usize] = cell;
                }
            }
            self.rebuild_lists();
            return true;
        }

        // Standard undo: remove piece from to, restore from and to cells
        self.cells[to] = EMPTY_CELL;
        self.cells[from] = undo.from_cell;
        if undo.to_cell != EMPTY_CELL {
            self.cells[to] = undo.to_cell;
        }

        // Restore mid-capture
        if undo.mid_sq != INVALID_SQ {
            self.cells[undo.mid_sq as usize] = undo.mid_cell;
        }

        // Restore range captures
        if let Some(ref caps) = undo.range_caps {
            for &(sq, cell) in caps {
                self.cells[sq as usize] = cell;
            }
        }

        // Rebuild piece lists (simpler than incremental update for correctness)
        self.rebuild_lists();
        true
    }

    fn rebuild_lists(&mut self) {
        self.piece_list_len = [0; 2];
        self.piece_count = [0; 2];
        self.royal_count = [0; 2];
        self.piece_list = [[INVALID_SQ; MAX_PIECES_PER_SIDE]; 2];
        self.royal_list = [[INVALID_SQ; MAX_ROYALS]; 2];

        for sq in 0..NUM_SQUARES {
            let cell = self.cells[sq];
            if cell != EMPTY_CELL {
                let pt = cell_piece(cell);
                let color = cell_color(cell);
                let c = color as usize;
                let idx = self.piece_list_len[c];
                self.piece_list[c][idx] = sq as u16;
                self.piece_list_len[c] = idx + 1;
                self.piece_count[c] += 1;
                if pieces::is_royal(pt) {
                    let ri = self.royal_count[c];
                    self.royal_list[c][ri] = sq as u16;
                    self.royal_count[c] = ri + 1;
                }
            }
        }
    }

    fn remove_from_lists(&mut self, sq: usize) {
        let cell = self.cells[sq];
        if cell == EMPTY_CELL { return; }
        let pt = cell_piece(cell);
        let color = cell_color(cell) as usize;
        self.remove_sq_from_piece_list(sq, color);
        self.piece_count[color] -= 1;
        if pieces::is_royal(pt) {
            self.remove_sq_from_royal_list(sq, color);
        }
    }

    fn remove_sq_from_piece_list(&mut self, sq: usize, color: usize) {
        let sq16 = sq as u16;
        let len = self.piece_list_len[color];
        for i in 0..len {
            if self.piece_list[color][i] == sq16 {
                self.piece_list[color][i] = self.piece_list[color][len - 1];
                self.piece_list[color][len - 1] = INVALID_SQ;
                self.piece_list_len[color] = len - 1;
                return;
            }
        }
    }

    fn remove_sq_from_royal_list(&mut self, sq: usize, color: usize) {
        let sq16 = sq as u16;
        let len = self.royal_count[color];
        for i in 0..len {
            if self.royal_list[color][i] == sq16 {
                self.royal_list[color][i] = self.royal_list[color][len - 1];
                self.royal_list[color][len - 1] = INVALID_SQ;
                self.royal_count[color] = len - 1;
                return;
            }
        }
    }

    fn add_sq_to_piece_list(&mut self, sq: usize, color: usize) {
        let idx = self.piece_list_len[color];
        self.piece_list[color][idx] = sq as u16;
        self.piece_list_len[color] = idx + 1;
        self.piece_count[color] += 1;
    }

    fn update_royal_status(&mut self, sq: usize, old_pt: u16, new_pt: u16, color: usize) {
        if pieces::is_royal(old_pt) && !pieces::is_royal(new_pt) {
            self.remove_sq_from_royal_list(sq, color);
        } else if !pieces::is_royal(old_pt) && pieces::is_royal(new_pt) {
            let ri = self.royal_count[color];
            self.royal_list[color][ri] = sq as u16;
            self.royal_count[color] = ri + 1;
        }
    }

    pub fn game_result(&self) -> Option<GameResult> {
        let b = self.royal_count[BLACK as usize] > 0;
        let w = self.royal_count[WHITE as usize] > 0;
        match (b, w) {
            (true, true) => None,
            (true, false) => Some(GameResult::BlackWins),
            (false, true) => Some(GameResult::WhiteWins),
            (false, false) => Some(GameResult::Draw),
        }
    }

    pub fn display(&self) -> String {
        let mut s = String::new();
        for r in 0..BOARD_SIZE {
            s.push_str(&format!("{:2} ", BOARD_SIZE - r));
            for c in 0..BOARD_SIZE {
                let cell = self.cells[sq_index(r, c)];
                if cell == EMPTY_CELL {
                    s.push_str(".. ");
                } else {
                    let pt = cell_piece(cell);
                    let color = cell_color(cell);
                    let prefix = if color == WHITE { 'v' } else { '^' };
                    let ab = pieces::abbrev(pt);
                    s.push(prefix);
                    s.push_str(&format!("{:<2}", &ab[..ab.len().min(2)]));
                }
            }
            s.push('\n');
        }
        s
    }
}
